src/Eloquent/Model.php

```diff
@@ -199,6 +202,33 @@ public function getAttribute($key)
         return parent::getAttribute($key);
     }
 
+    /** @inheritdoc */
+    protected function transformModelValue($key, $value)
+    {
+        $value = parent::transformModelValue($key, $value);
+        if ($this->hasCast($key) && $value instanceof CarbonInterface) {
```

Comment:
I used `CarbonInterface` to cover both `Carbon` and `CarbonImmutable` instances.